MWI: Add documentation for bound keypair joining

timothyb89 opened the pull request:
This adds documentation for bound keypair joining, including a user-facing deployment guide, an admin guide (with architecture information), and the usual provision token details.

Closes #55598

timothyb89 committed:
MWI: Add documentation for bound keypair joining

This adds documentation for bound keypair joining, including a
user-facing deployment guide, an admin guide (with architecture
information), and the usual provision token details.

timothyb89 committed:
TODOs for deployment guide, add deployment guide to method list

strideynet commented on docs/pages/reference/machine-id/bound-keypair.mdx:
I think pending the wider docs refactor - this is in the perfect place and I think strikes a nice level of detail.

timothyb89 committed:
Finish deployment guide and make small tweaks to other pages

timothyb89 committed:
Update tbot CLI and configuration reference

timothyb89 committed:
Fix docs lints

ptgott commented on docs/pages/reference/machine-id/bound-keypair.mdx:
```diff
@@ -0,0 +1,567 @@
+  use cases and deployment scenarios
+- Ensures failed bots can be recovered without client-side intervention in most
+  cases
+
```
I would be clearer about what readers would expect in this guide. The introduction frames it as a general-purpose docs page on bound keypair joining, but that puts it in conflict with the how-to guide. I would consider:
- Splitting the architectural information into a separate guide with a more specific title and putting that in the Architecture section.
- Retitling this page to be more specific, e.g., Bound Keypair Joining Reference.
- Making sure each Bound Keypair Joining guide links to the other Bound Keypair Joining guides in the introduction.

zmb3 replied:
I like option 2.

timothyb89 replied:
I've at least retitled the page and added a guides section to link to the main deployment guide. I'll admit it definitely doesn't fit cleanly into "just" being a reference page and is sort of a cursed hybrid of an architecture page, admin manual, and configuration reference.

I'm not opposed making additional separate pages, but I think (for example) a dedicated architecture page might be a bit light on content given the "Concepts" section as it exists now. Definitely happy to hear your thoughts on the best path forward!

zmb3 commented on docs/pages/machine-workload-identity/machine-id/deployment/bound-keypair.mdx:
```diff
@@ -0,0 +1,174 @@
+---
+
+In this guide, you will install Machine ID's agent, `tbot`, on an arbitrary
+node. This node could be a bare-metal machine, a VM, a container, or any other
```
```suggestion
host. This node could be a bare-metal machine, a VM, a container, or any other
```

"node" pretty consistently means "SSH server" in the Teleport docs, so let's avoid that confusion.

zmb3 commented on docs/pages/machine-workload-identity/machine-id/deployment/bound-keypair.mdx:
```diff
@@ -0,0 +1,174 @@
+description: "How to install and configure Machine ID with Bound Keypair joining"
+---
+
+In this guide, you will install Machine ID's agent, `tbot`, on an arbitrary
```
+1 for the first option - we should explain what this join method is and when you should consider using it as early as we can.

timothyb89 replied:
I've tried to wordsmith things here to hopefully cover everything: it directly compares against other options and mentions recovery.

zmb3 commented on docs/pages/machine-workload-identity/machine-id/deployment/bound-keypair.mdx:
```diff
@@ -0,0 +1,174 @@
+
+## How it works
+
+With Bound Keypair joining, instead of using a static token or relying solely on
+renewable certificates, Machine ID bots generate a unique keypair. Teleport is
+then configured to trust this public key for future joining attempts.
```
I would intro it a bit differently. Take it or leave it:

> Bound keypair joining is a more secure and flexible alternative to static token joining.
> Machine ID bots generate a unique keypair, registering the public key with Teleport
> and keeping the private key in their own storage.
>
> Later, when the bot attempts to join the cluster, ...

timothyb89 replied:
I've tried to iterate on this a bit and rewrote things a bit, the "more secure and flexible alternative" wound up getting promoted up into the intro paragraph

zmb3 commented on docs/pages/machine-workload-identity/machine-id/deployment/bound-keypair.mdx:
```diff
@@ -0,0 +1,174 @@
+  duplicated
+- Certificate generation counter checks to ensure regular bot certificates
+  cannot be usefully shared or duplicated
+- Configurable limits on how often - if at all - bots may be allowed to recover
```
Do we expect readers to know what it means to "recover" at this point in the guide, or do we need to elaborate?

zmb3 commented on docs/pages/machine-workload-identity/machine-id/deployment/bound-keypair.mdx:
```diff
@@ -0,0 +1,174 @@
+
+**This step is completed on your local machine.**
+
+In this guide, we'll demonstrate joining a bot using a registration secret: this
```
I would mention both approaches before we narrow down on one, just so the reader knows there's another option.

Maybe:

> The initial joining operation can operate in two modes. The first is by using
> a registration secret, which is similar to the first join process with the `token`
> join method. Alternatively, initial joining can be performed with a public key.
>
> In this guide, we will demonstrate joining using a registration secret.

timothyb89 replied:
I've added an explicit reference (and a link) in the admonition, hopefully it makes the existence of the alternative clear? Trying to balance clarity with the guide only actually showing how to accomplish a single mode.

(Maybe showing both paths in the guide will be more reasonable with a `tctl` helper - may update this further if so)

zmb3 commented on docs/pages/machine-workload-identity/machine-id/deployment/bound-keypair.mdx:
````diff
@@ -0,0 +1,174 @@
+      limit: 1
+```
+
+Replace:
+- `example` in `spec.bot_name` with the name of the bot you created in the
+  second step
````
I would just write this as a plain english sentence. A list containing one item looks a little awkward to me.

```suggestion
Replace `example` in `spec.bot_name` with the name of the bot you created in the second step.
```

zmb3 commented on docs/pages/machine-workload-identity/machine-id/deployment/bound-keypair.mdx:
````diff
@@ -0,0 +1,174 @@
+Next, retrieve the generated registration secret, which will be needed for the
+next step:
+```code
+$ tctl get token/example --format=json | jq -r '.[0].status.bound_keypair.registration_secret'
````
This flow might be much easier to document (and user friendly) if we had a `tctl tokens add` command which created the token and output the registration secret.

timothyb89 replied:
Hrm, I agree. I'll see if I can't throw something together.

zmb3 commented on docs/pages/reference/join-methods.mdx:
```diff
@@ -264,6 +264,19 @@ When a MachineID bot uses an ephemeral join token, the token is deleted.
+### Bound Keypair: `bound_keypair`
+
+Bound Keypair tokens are an improved alternative to
+[ephemeral tokens](#ephemeral-tokens) that aims to improve reliability for
```
```suggestion
[ephemeral tokens](#ephemeral-tokens) that aim to improve reliability for
```

A token aims.
Multiple tokens aim.

timothyb89 replied:
Ack, whoops, good catch!

zmb3 commented on docs/pages/reference/machine-id/bound-keypair.mdx:
```diff
@@ -0,0 +1,567 @@
+---
+
+Bound Keypair is a new join method designed to provide the best features of
+[dynamic join methods][dynamic] join methods - like AWS, GCP, or Azure - but in
```
Should we call them _delegated join methods_ for consistency with https://goteleport.com/docs/reference/join-methods/#secret-vs-delegated

zmb3 commented on docs/pages/reference/machine-id/bound-keypair.mdx:
```diff
@@ -0,0 +1,567 @@
+supported by a delegated join method.
+
+As a drop-in replacement for `token` joining, Bound Keypair joining is also a
+good replacement for local experimentation for testing, and can be easily
```
can be easily what?

zmb3 commented on docs/pages/reference/machine-id/bound-keypair.mdx:
````diff
@@ -0,0 +1,567 @@
+method is to use a join token lock target:
+
+```code
+$ tctl lock --join-token=token-name
````
Does this work for _any_ join token, or does it have to be a bound_keypair join token?

timothyb89 replied:
It was added for bound keypair use, but will work with any non-`token` token. Looks like I forgot to include these new targets in the locking docs, so I'll add a reference.

timothyb89 committed:
Add notes about new lock targets for bots

timothyb89 committed:
Bound keypair guide feedback

Various bits of review feedback, and attempt to wordsmith on the
introduction paragraphs.

timothyb89 committed:
Fix review feedback and missing references

Adds some missing references and addresses review feedback.

timothyb89 committed:
Address review feedback, add missing detail

Adds missing detail, especially around decoupled keypair from certs.

timothyb89 committed:
Further clarify what automatic recovery is

timothyb89 committed:
Consistently refer to "bot hosts" rather than nodes

timothyb89 committed:
Address more review feedback

- Retitle the reference guide
- Link to preregistered keys in the guide

timothyb89 committed:
Add guides section to reference page

timothyb89 committed:
Remove admin guide from this PR

This removes the admin guide from this PR and replaces references to
it with placeholders with commented TODOs. The admin guide will be
reintroduced in a separate PR stacked onto this one.

timothyb89 committed:
Fix docs lint

timothyb89 committed:
Fix Machine ID -> Machine & Workload Identity naming

timothyb89 committed:
Update more Machine & Workload Identity references
